### PR TITLE
Deprecated Various Component Methods from `Query` and `QueryState`

### LINKED ISSUE #9904: ``Query::get_component`` does not respect filters
## Bevy version

0.11.2

## What you did

Created a system query: ``query: Query<&Transform, Changed<Transform>>`` and then used ``query.get_component(entity_with_unchanging_transform)``.

## What went wrong

- what were you expecting?
I expected to get no result since the query has the Changed filter.
- what actually happened?
I got a result.

## Additional information

- theories about what might be going wrong
 From looking at the source it seems like the ``Query::get_component[_mut]()`` variants are missing logic to respect the filter of the query.
 Or the name of the method fails to clarify its use. Perhaps ``get_component_ignore_filter()`` is more suitable if this is the intended behaviour.

- links to related bugs, PRs or discussions
https://github.com/bevyengine/bevy/pull/9523
https://github.com/bevyengine/bevy/issues/9011

### PR DESCRIPTION
**Pull request by bushrat011899**
# Objective

- (Partially) Fixes #9904
- Acts on #9910

## Solution

- Deprecated the relevant methods from `Query`, cascading changes as required across Bevy.

---

## Changelog

- Deprecated `QueryState::get_component_unchecked_mut` method
- Deprecated `Query::get_component` method
- Deprecated `Query::get_component_mut` method
- Deprecated `Query::component` method
- Deprecated `Query::component_mut` method
- Deprecated `Query::get_component_unchecked_mut` method

### `QueryState::get_component_unchecked_mut`

Use `QueryState::get_unchecked_manual` and select for the exact component based on the structure of the exact query as required.

### `Query::(get_)component(_unchecked)(_mut)`

Use `Query::get` and select for the exact component based on the structure of the exact query as required.

- For mutable access (`_mut`), use `Query::get_mut`
- For unchecked access (`_unchecked`), use `Query::get_unchecked`
- For panic variants (non-`get_`), add `.unwrap()`

## Notes

- `QueryComponentError` can be removed once these deprecated methods are also removed. Due to an interaction with `thiserror`'s derive macro, it is not marked as deprecated.

### CODE REVIEW

File: benches/benches/bevy_ecs/world/world_get.rs [hunk 1]
```diff
@@ -319,7 +319,7 @@ pub fn query_get_component(criterion: &mut Criterion) {
                 let mut count = 0;
                 for comp in entities
                     .iter()
-                    .flat_map(|&e| query.get_component::<Table>(e))
```
Comment (james7132) [inline benches/benches/bevy_ecs/world/world_get.rs:322]: This entire benchmark is focused on get_component, so we should probably just remove the entire thing.

File: crates/bevy_ecs/src/system/query.rs [hunk 2]
```diff
@@ -1,16 +1,15 @@
...
+        BatchingStrategy, QueryCombinationIter, QueryEntityError, QueryIter, QueryManyIter,
+        QueryParIter, QuerySingleError, QueryState, ROQueryItem, ReadOnlyWorldQuery, WorldQuery,
     },
-    world::{unsafe_world_cell::UnsafeWorldCell, Mut},
+    world::unsafe_world_cell::UnsafeWorldCell,
 };
-use std::{any::TypeId, borrow::Borrow};
+use std::borrow::Borrow;
 
-/// [System parameter] that provides selective access to the [`Component`] data stored in a [`World`].
```
Comment (james7132) [inline crates/bevy_ecs/src/system/query.rs:13]: Revert this and keep the backticks for formatting purposes. The ref-link can use the backticks too.

**Review by james7132 — COMMENTED** [review]
Small comments here but happy with the overall change. Less confusion and less perf footguns.

Please remove the pub(crate) and private parts from the changelog in the PR description. The changelog and migration guide are supposed to be user facing.

**Comment by james7132** [comment]
With that said, IMO this final decision should be left to @cart, even with SME approval. This is one of the core functions on Query, and it's one of the primary parts of Bevy's query ergonomics. The poll in #9910 currently is mostly in favor of removal, but I do also want to get a more representative sample from the community before moving forward with this.

File: benches/benches/bevy_ecs/world/world_get.rs (hunk 1 above)
Comment (bushrat011899) [reply benches/benches/bevy_ecs/world/world_get.rs:322]: Fair enough, I'll remove that test!

File: crates/bevy_ecs/src/system/query.rs (hunk 2 above)
Comment (bushrat011899) [reply crates/bevy_ecs/src/system/query.rs:13]: My bad on the documentation! Fixed up.

**Comment by bushrat011899** [comment]
> With that said, IMO this final decision should be left to @cart [...] I do also want to get a more representative sample from the community before moving forward with this.
Totally reasonable, I thought I'd just mock up a PR for what this change would represent to help with the decision making process.

**Comment by james7132** [comment]
At @nicopap's suggestion and the poll's results, it does seem like deprecation and eventual removal is the way to go here. If this makes it into 0.12, the deprecation can act as a shout test for us before we remove it completely in a subsequent release.

**Review by cart — APPROVED** [review]
I'm on board for this. `get_component` is rarely used, less efficient, a safety hazard, and a maintenance risk. If we're going to eat the costs of this, it should be more useful than it currently is. 

I _do_ think we have a "query abstraction" problem that makes it hard to write functionality that is reusable across queries. There is also the "needle in the haystack" problem where finding the right component in a large query is ugly / hard:

```rust
// This is nasty to write, nasty to read, and nasty adjust as the query changes
let (_, _, _, _, foo, _, _, _) = query.get(entity).unwrap();

// This is easier on the eyes but forces users to manually count
// (both when writing and reading)
// Still nasty
let foo = query.get(entity).unwrap().4;
```

I think theres space for something like a "query view" system (which would internally look a bit like `get_component`) that makes it possible to use any query that yields the required components. I think solving that problem would likely also solve the problems that `get_component` aims to solve.

**Comment by cart** [comment]
I'm down to deprecate in 0.12 and then remove in 0.13

**Review by james7132 — APPROVED** [review]
If we can change this PR for deprecation instead of removal, I'd like to get this in before 0.13 lands so we can move the removal timeline up.

**Comment by bushrat011899** [comment]
> If we can change this PR for deprecation instead of removal, I'd like to get this in before 0.13 lands so we can move the removal timeline up.

I've left all the methods to be removed as-is, instead marking as deprecated. Because I was having an issue with Clippy I couldn't mark `QueryComponentError` as deprecated as well. I suspect the `thiserror` `Error` derive macro expands in such a way that the deprecation still triggers a warning even when used with `allow(deprecated)`.

**Comment by nicopap** [comment]
In my humble opinion, query transmutes definitively make those component methods out-of-date.

**Review by nicopap — APPROVED** [review]

**Review by james7132 — APPROVED** [review]

**Comment by rparrett** [comment]
IMO, this migration guide could use a code block illustrating what is meant by 
> and select for the exact component based on the structure of the exact query as required

No need to go crazy or cover every case, but something like this would have saved me a minute of digging up this PR and figuring out what was going on.

```rust
fn system(query: Query<(&A, &B, &C)>) {
    // Before
    let b = query.get_component::<B>(entity).unwrap();
    // After
    let (_, b, _) = query.get(entity).unwrap();
}
```

A short passage explaining the motivation for the breakage can be helpful too for a users' sanity.

> These methods were rarely used, less efficient than the alternatives, and a burden on Bevy maintainers, so they have been deprecated.

Without this sentence, I'm slightly grumpy because this seems incredibly arbitrary. With this sentence, I am actually happy to be making this change to my code.

**Comment by bushrat011899** [comment]
> IMO, this migration guide could use a code block illustrating what is meant by
> 
> > and select for the exact component based on the structure of the exact query as required
> 
> No need to go crazy or cover every case, but something like this would have saved me a minute of digging up this PR and figuring out what was going on.
> 
> ```rust
> fn system(query: Query<(&A, &B, &C)>) {
>     // Before
>     let b = query.get_component::<B>(entity).unwrap();
>     // After
>     let (_, b, _) = query.get(entity).unwrap();
> }
> ```

Totally fair, I've updated the migration guide based on this code example, also including a second alternative which may be preferred by some users.

> A short passage explaining the motivation for the breakage can be helpful too for a users' sanity.
> 
> > These methods were rarely used, less efficient than the alternatives, and a burden on Bevy maintainers, so they have been deprecated.
> 
> Without this sentence, I'm slightly grumpy because this seems incredibly arbitrary. With this sentence, I am actually happy to be making this change to my code.

Also completely fair, I've added the suggested sentence verbatim, with an additional note to refer to the linked issue and discussion in a new section labelled `Motivation`.

Thanks for sharing your experience, hopefully this helps others along the way!

**Comment by rparrett** [comment]
Thanks! Not a huge deal in the grand scheme of things, but I really appreciate any help with the migration guides I can get!

**Comment by haath** [comment]
So after the removal, the only way now to get a specific component by its type is to use [Query transmutation](https://bevyengine.org/news/bevy-0-13/#query-transmutation), correct?

And to be clear, I am referring to use cases where you don't know the exact structure of the query, like in the following example which I have in `0.12`:

```rust
// with 0.12
fn get_child_component<'a, C, Q, F>(children: &Children, self, query: &'a Query<'_, '_, Q, F>) -> Option<&'a C>
where
    C: Component,
    Q: WorldQuery,
    F: ReadOnlyWorldQuery,
{
    for &child in children
    {
        if let Ok(component) = query.get_component::<C>(child)
        {
            return Some(component);
        }
    }

    None
}
```

If I understand correctly, starting with `0.13` this function will need to be rewritten to accept a `QueryLens`.

However the release notes also mention the following:

> One thing to take into consideration is the transmutation is not free. It works by creating a new state and copying a bunch of the cached data inside the original query. It's not a expensive operation, but you should probably avoid doing it inside a hot loop.

So does this mean that the removal of these methods now imposes a performance penalty on these use cases?

**Pull request merged**